File: terrareg/provider_source/repository_release_metadata.py

from typing import Union, List

import re


class RepositoryReleaseMetadata:
    """Structure for storing repository release metadata"""

    _TAG_REGEX = re.compile(r'^v([0-9]+\.[0-9]+\.[0-9]+(:?-[a-z0-9]+)?)$')

    def __init__(self, name: str, tag: str, archive_url: str, commit_hash: str,
                 provider_id: Union[str, int], release_artifacts: List['ReleaseArtifactMetadata']):
        """Store member variables"""
        self.name = name
        self.tag = tag
        self.provider_id = provider_id
        self.release_artifacts = release_artifacts
        self.archive_url = archive_url
        self.commit_hash = commit_hash

    def __eq__(self, __o):
        """Check if two repository releases are the same"""
        if isinstance(__o, self.__class__):
            return (
                self.name == __o.name and
                self.tag == __o.tag and
                self.provider_id == __o.provider_id and
                self.release_artifacts == __o.release_artifacts and
                self.archive_url == __o.archive_url and
                self.commit_hash == __o.commit_hash
            )
        return super(RepositoryReleaseMetadata, self).__eq__(__o)

    @classmethod
    def tag_to_version(cls, tag: str) -> Union[None, str]:
        """Convert tag to version"""
        # Since we currently support the original tagging strategy 'v{version}', simply strip off the v
        if (match := cls._TAG_REGEX.match(tag)):
            return match.group(1)
        return None

    @property
    def version(self) -> Union[None, str]:
        """Convert tag to version"""
        return self.tag_to_version(self.tag)


class ReleaseArtifactMetadata:
    """Structure for storing release artifact metadata"""

    def __init__(self, name: str, provider_id: Union[str, int]):
        """Store member variables"""
        self.name = name
        self.provider_id = provider_id

    def __eq__(self, __o):
        """Check if two release artifacts are the same"""
        if isinstance(__o, self.__class__):
            return self.name == __o.name and self.provider_id == __o.provider_id
        return super(ReleaseArtifactMetadata, self).__eq__(__o)

File: terrareg-go/internal/domain/provider_source/repository_release_metadata.go
package provider_source

import (
	"regexp"
)

// tagRegex matches version tags like v1.5.2, v1.5.2-beta1, etc.
var tagRegex = regexp.MustCompile(`^v([0-9]+\.[0-9]+\.[0-9]+(-[a-z0-9]+)?)$`)

// ReleaseArtifactMetadata stores release artifact metadata
// Python reference: provider_source/repository_release_metadata.py::ReleaseArtifactMetadata
type ReleaseArtifactMetadata struct {
	Name       string
	ProviderID string
}

// NewReleaseArtifactMetadata creates a new ReleaseArtifactMetadata
func NewReleaseArtifactMetadata(name, providerID string) *ReleaseArtifactMetadata {
	return &ReleaseArtifactMetadata{
		Name:       name,
		ProviderID: providerID,
	}
}

// Equals checks if two release artifacts are the same
// Python reference: ReleaseArtifactMetadata.__eq__
func (r *ReleaseArtifactMetadata) Equals(other *ReleaseArtifactMetadata) bool {
	if other == nil {
		return false
	}
	return r.Name == other.Name && r.ProviderID == other.ProviderID
}

// RepositoryReleaseMetadata stores repository release metadata
// Python reference: provider_source/repository_release_metadata.py::RepositoryReleaseMetadata
type RepositoryReleaseMetadata struct {
	Name            string
	Tag             string
	ArchiveURL      string
	CommitHash      string
	ProviderID      string
	ReleaseArtifacts []*ReleaseArtifactMetadata
}

// NewRepositoryReleaseMetadata creates a new RepositoryReleaseMetadata
func NewRepositoryReleaseMetadata(
	name, tag, archiveURL, commitHash, providerID string,
	releaseArtifacts []*ReleaseArtifactMetadata,
) *RepositoryReleaseMetadata {
	return &RepositoryReleaseMetadata{
		Name:            name,
		Tag:             tag,
		ArchiveURL:      archiveURL,
		CommitHash:      commitHash,
		ProviderID:      providerID,
		ReleaseArtifacts: releaseArtifacts,
	}
}

// Equals checks if two repository releases are the same
// Python reference: RepositoryReleaseMetadata.__eq__
func (r *RepositoryReleaseMetadata) Equals(other *RepositoryReleaseMetadata) bool {
	if other == nil {
		return false
	}
	if r.Name != other.Name ||
		r.Tag != other.Tag ||
		r.ProviderID != other.ProviderID ||
		r.ArchiveURL != other.ArchiveURL ||
		r.CommitHash != other.CommitHash ||
		len(r.ReleaseArtifacts) != len(other.ReleaseArtifacts) {
		return false
	}

	// Compare release artifacts
	for i, artifact := range r.ReleaseArtifacts {
		if !artifact.Equals(other.ReleaseArtifacts[i]) {
			return false
		}
	}

	return true
}

// TagToVersion converts a tag to a version string
// Returns nil if the tag doesn't match the expected format
// Python reference: RepositoryReleaseMetadata.tag_to_version
// Examples: "v1.5.2" -> "1.5.2", "v1.5.2-beta1" -> "1.5.2-beta1", "invalid" -> nil
func TagToVersion(tag string) *string {
	matches := tagRegex.FindStringSubmatch(tag)
	if len(matches) < 2 {
		return nil
	}
	version := matches[1]
	return &version
}

// VersionToTag converts a version string to a tag
// Python reference: RepositoryReleaseMetadata.version_to_tag
// Examples: "1.5.2" -> "v1.5.2", "1.5.2-beta1" -> "v1.5.2-beta1"
func VersionToTag(version string) string {
	return "v" + version
}

// Version returns the version string from the tag
// Returns nil if the tag doesn't match the expected format
// Python reference: RepositoryReleaseMetadata.version
func (r *RepositoryReleaseMetadata) Version() *string {
	return TagToVersion(r.Tag)
}
